### PR TITLE
DIT-2349 changes to navigation

### PR DESCRIPTION
**Pull request by DavidGregory084**
This contains the navigation changes from DIT-2349 that Adam and Gorkem have worked on. I think I have updated the rest of the journey and the check your answers navigation.

Most of our controllers now do the same thing. I have captured this in [QuestionCachingController](https://github.com/hmrc/binding-tariff-trader-frontend/compare/DIT-2349?expand=1#diff-9a2789eb46c2be9b0d9810814a465f55).

There are a few major exceptions to this in the journey:

* BeforeYouStartController - this controller creates the initial UserAnswers and sets it into the session
* The file upload / file list pages - they are a bit complex at the moment - I am still considering how to handle that, perhaps by adding an extra page for the first trip through "do you want to add supporting documents"?
* Check your answers / declaration controller - that part is the end of the journey obviously so doesn't fit into the QuestionPage abstraction

### CODE REVIEW

File: app/controllers/YesNoBehaviour.scala [hunk 1]
```diff
@@ -18,32 +18,39 @@ package controllers
...
+        case (userAnswers, detailPage) => userAnswers.remove(detailPage)
+      }
+
+    dataCacheConnector.save(updatedAnswers.cacheMap).transformWith {
+      case Failure(NonFatal(_)) =>
+        Future.successful(Results.BadGateway)
+      case Success(_) =>
+        Future.successful(Results.Redirect(navigator.nextPage(questionPage, mode)(updatedAnswers)))
+    }
   }
```
Comment (DavidGregory084) [inline app/controllers/YesNoBehaviour.scala:54]: Note to self: this file needs to be renamed

File: test/unit/navigation/NavigatorSpec.scala [hunk 2]
```diff
@@ -80,13 +89,209 @@ class NavigatorSpec extends SpecBase {
...
+
+        navigator.nextPage(SelectApplicationTypePage, NormalMode)(mockUserAnswers) shouldBe
+          routes.SelectApplicationTypeController.onPageLoad(NormalMode)
+
+        navigator.nextPage(SimilarItemCommodityCodePage, NormalMode)(mockUserAnswers) shouldBe
+          routes.SimilarItemCommodityCodeController.onPageLoad(NormalMode)
+      }
     }
 
     "in Check mode" must {
```
Comment (OC-Pomaroff) [inline test/unit/navigation/NavigatorSpec.scala:289]: Is it worthwhile adding tests here for the yes/no journeys in checkmode?

**Review by OC-Pomaroff — APPROVED** [review]
Looking awesome!

File: test/unit/navigation/NavigatorSpec.scala (hunk 2 above)
Comment (DavidGregory084) [reply test/unit/navigation/NavigatorSpec.scala:289]: Yes in fact that was the main thing I was trying to achieve today before I got sidetracked 😂

File: app/controllers/SelectApplicationTypeController.scala [hunk 3]
```diff
@@ -21,60 +21,33 @@ import connectors.DataCacheConnector
...
+  val dataCacheConnector: DataCacheConnector,
+  val navigator: Navigator,
+  val identify: IdentifierAction,
+  val getData: DataRetrievalAction,
+  val requireData: DataRequiredAction,
+  formProvider: SelectApplicationTypeFormProvider,
+  cc: MessagesControllerComponents
+)(implicit ec: ExecutionContext) extends YesNoCachingController(cc) {
+  lazy val form = formProvider()
+  val journey = Journey.previousBTI
```
[the remark this answers is not in the document]
Comment (OC-Pomaroff) [reply app/controllers/SelectApplicationTypeController.scala:47]: haha, sure 🤣

**Review by kuldeepwadhwa9 — APPROVED** [review]

**Pull request merged**